### ISSUE TITLE
Reload R model object imported from different environment without using a temp file

### ISSUE THREAD

**Issue by JesseLimtiaco** [Microsoft/LightGBM#432]
Hello,

My question is similar to the issue brought up in 296:

https://github.com/Microsoft/LightGBM/issues/296

The R6 objects are environment-type variables and my use case transfers these R6 objects between different environments/machines instead of RDS files.  The solution to make the saveRDS.lgb.Booster and readRDS.lgb.Booster works when transferring the RDS files but I was hoping to avoid creating a temp file in the R session.  

readRDS.lgb.Booster looks like this where it reads off the temp file that is written to file storage.  

readRDS.lgb.Booster <- function (file = "", refhook = NULL) 
{
  object <- readRDS(file = file, refhook = refhook)
  if (!is.na(object$raw)) {
    temp <- tempfile()
    write(object$raw, temp)
    object2 <- lgb.load(temp)
    file.remove(temp)
    object2$best_iter <- object$best_iter
    object2$record_evals <- object$record_evals
    return(object2)
  }
  else {
    return(object)
  }
}

Can we make something that looks like this that passes in an R6 object that already has a saved $raw attribute?

reload.lgb.Booster <- function (object) 
{
  if (!is.na(object$raw)) {
    object2 <- lgb.load(object$raw) # initializes new model from object$raw data instead of text file
    object2$best_iter <- object$best_iter
    object2$record_evals <- object$record_evals
    return(object2)
  }
  else {
    return(object)
  }
}

Thanks!

**Comment by Laurae2** [comment]
Can you save the model and the best iteration / metric log separately? If yes, then this will allow you to skip this issue. You can then write the metric log / best iteration directly on the newly created environment (model).

ping @guolinke if there is a way to save/load model from a variable containing the model string in memory instead of a file ?

**Comment by guolinke** [comment]
@Laurae2 
https://github.com/Microsoft/LightGBM/blob/master/src/c_api.cpp#L1035-L1048
https://github.com/Microsoft/LightGBM/blob/master/src/c_api.cpp#L769-L779

**Comment by JesseLimtiaco** [comment]
Yes this implementation would be very helpful.  Can we add analogous __load_model_from_string and __save_model_to_string functions to the Booster class in the R package?  Right now it seems the only way to initialize in R is to make a new model or read from model file:

# Booster non existent
        stop("lgb.Booster: Need at least either training dataset or model file to create booster instance")

**Comment by JesseLimtiaco** [comment]
Solved with pull request: Adding ability to load model from string in R #472
